File: leetcode/3336-find-the-number-of-subsequences-with-equal-gcd/3336-find-the-number-of-subsequences-with-equal-gcd.ts
function subsequencePairCount(nums: number[]): number {
    const MOD: number = 1_000_000_007;
    const MAX: number = 201;
    const SIZE: number = MAX * MAX;

    function gcd(a: number, b: number): number {
        while (b !== 0) {
            const temp: number = b;
            b = a % b;
            a = temp;
        }

        return a;
    }

    // GCD 미리 계산
    const gcdTable: Int16Array[] = Array.from(
        { length: MAX },
        (): Int16Array => new Int16Array(MAX)
    );

    for (let i = 0; i < MAX; i++) {
        for (let j = 0; j < MAX; j++) {
            gcdTable[i][j] = gcd(i, j);
        }
    }

    let dp: Int32Array = new Int32Array(SIZE);

    // dp[0][0] = 1
    dp[0] = 1;

    for (const num of nums) {
        const next: Int32Array = new Int32Array(SIZE);

        for (let g1 = 0; g1 < MAX; g1++) {
            const base: number = g1 * MAX;

            for (let g2 = 0; g2 < MAX; g2++) {
                const index: number = base + g2;
                const count: number = dp[index];

                if (count === 0) {
                    continue;
                }

                // 1. 현재 숫자를 선택하지 않음
                let value: number = next[index] + count;

                if (value >= MOD) {
                    value -= MOD;
                }

                next[index] = value;

                // 2. 첫 번째 부분 수열에 추가
                const nextG1: number = gcdTable[g1][num];
                const index1: number = nextG1 * MAX + g2;

                value = next[index1] + count;

                if (value >= MOD) {
                    value -= MOD;
                }

                next[index1] = value;

                // 3. 두 번째 부분 수열에 추가
                const nextG2: number = gcdTable[g2][num];
                const index2: number = base + nextG2;

                value = next[index2] + count;

                if (value >= MOD) {
                    value -= MOD;
                }

                next[index2] = value;
            }
        }

        dp = next;
    }

    let answer: number = 0;

    for (let g = 1; g < MAX; g++) {
        answer += dp[g * MAX + g];

        if (answer >= MOD) {
            answer -= MOD;
        }
    }

    return answer;
}

File: leetcode/3336-find-the-number-of-subsequences-with-equal-gcd/3336-find-the-number-of-subsequences-with-equal-gcd.java
import java.util.Arrays;

class Solution {
    private static final int MOD = 1_000_000_007;

    // mem[i][gcd1][gcd2]
    private int[][][] mem;
    private int[] nums;

    private int solve(int i, int g1, int g2) {
        // 모든 원소를 확인한 경우
        if (i == nums.length) {
            // 두 부분 수열의 GCD가 같은 경우
            return g1 > 0 && g1 == g2 ? 1 : 0;
        }

        // 이미 계산한 상태라면 메모이제이션된 값 반환
        if (mem[i][g1][g2] != -1) {
            return mem[i][g1][g2];
        }

        // 선택지 1: 현재 원소를 어느 부분 수열에도 포함하지 않음
        long ans = solve(i + 1, g1, g2);

        // 선택지 2: 현재 원소를 첫 번째 부분 수열에 포함
        int nextG1 = g1 == 0
                ? nums[i]
                : gcd(g1, nums[i]);

        ans = (ans + solve(i + 1, nextG1, g2)) % MOD;

        // 선택지 3: 현재 원소를 두 번째 부분 수열에 포함
        int nextG2 = g2 == 0
                ? nums[i]
                : gcd(g2, nums[i]);

        ans = (ans + solve(i + 1, g1, nextG2)) % MOD;

        return mem[i][g1][g2] = (int) ans;
    }

    private int gcd(int a, int b) {
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }

        return a;
    }

    public int subsequencePairCount(int[] nums) {
        this.nums = nums;

        mem = new int[nums.length][201][201];

        // 메모이제이션 배열 초기화
        for (int i = 0; i < nums.length; i++) {
            for (int j = 0; j <= 200; j++) {
                Arrays.fill(mem[i][j], -1);
            }
        }

        return solve(0, 0, 0);
    }
}